[DEPRECATION canary] Deprecate Ember.String and prototype extension

Serabe opened the pull request:
This PR deprecates the use of String prototype extension and
Ember.String namespace.

Methods coming from ember-runtime
---

A new `StringUtils` object is added, to be used internally by Ember
instead of using the same object being exposed as `Ember.String`. This
simplifies testing and the changes required in the rest of the code.
Also, dropping the deprecated code will be much simpler.

Methods coming from ember-glimmer
---

Deprecated versions are being added and reexported as the same global
the non-deprecated versions were. Besides, the non-deprecated version is
being exported under the new global `Ember.Template`.

The `Ember.Template` is just a placeholder until the final global is chosen.

Serabe commented:
@locks any update on this?

Serabe commented:
restarting job. Seemed a problem with yarn.

Serabe commented:
Passing.

rwjblue commented on packages/ember-glimmer/lib/utils/string.ts:
```diff
@@ -98,6 +99,19 @@ export function htmlSafe(str: string) {
+    {
+      id: 'ember-glimmer.ember-string-html-safe',
+      until: '3.5.0',
+      url: ''
```
Do we know the URL's for these deprecations? I vaguely recall that there was already an open website PR for them...

Serabe replied:
Adding the URL from https://github.com/emberjs/website/pull/2995/files

rwjblue commented on packages/ember-runtime/lib/ext/string.js:
```diff
@@ -14,9 +14,27 @@ import {
 const StringPrototype = String.prototype;
 
+function deprecateEmberStringExtension(fn, opts = {}) {
+  let { name } = fn;
```
I don't think we can rely on `Function.name` (AFAIK it isn't supported on all platforms, which would make this fail in production).

Checkout https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/name for a decent compat table (it isn't supported on IE at all)...

rwjblue commented on packages/ember-runtime/lib/system/string.js:
```diff
@@ -115,6 +115,22 @@ function capitalize(str) {
   return CAPITALIZE_CACHE.get(str);
 }
 
+function deprecateEmberStringUtil(fn, opts = {}) {
```
This seems roughly like the same function in the other file, can we share between the two and only have one implementation?

rwjblue commented on packages/ember-runtime/lib/system/string.js:
```diff
@@ -115,6 +115,22 @@ function capitalize(str) {
 }
 
+function deprecateEmberStringUtil(fn, opts = {}) {
+  let { name } = fn;
```
Same concern as above RE: `Function.name`

rwjblue commented on packages/ember/lib/index.js:
```diff
@@ -516,6 +527,19 @@ Object.defineProperty(Ember, 'TEMPLATES', {
+ @type Object
+ @private
+ */
+Ember.Template = {
```
Lets underscore this, I'd hate for folks to start using it ....

rwjblue reviewed (commented):
Looking really good!  Only a few small questions/adjustments that I commented on inline...

Serabe commented:
Addressed your suggestions. Thanks for the review!

Serabe commented:
Rebased and squashed.

Serabe commented:
Having lots of timeouts while doing `yarn install` in CI

Serabe commented:
Yay! I got tests passing after restarting some jobs several times.

mmun commented:
@rwjblue should we merge this?

locks commented:
@mmun negatory, sir. we need to publish the deprecation guides and release the addon before.
sorry for dwelling on this 😞

Serabe commented:
Rebased!

Serabe commented:
I will rebase this at some point, quite likely, next Friday. Would that be ok, @locks ?

Serabe committed:
[DEPRECATION canary] Deprecate Ember.String and prototype extension

This PR deprecates the use of String prototype extension and
Ember.String namespace.

Methods coming from ember-runtime
---

A new `StringUtils` object is added, to be used internally by Ember
instead of using the same object being exposed as `Ember.String`. This
simplifies testing and the changes required in the rest of the code.
Also, dropping the deprecated code will be much simpler.

Methods coming from ember-glimmer
---

Deprecated versions are being added and reexported as the same global
the non-deprecated versions were. Besides, the non-deprecated version is
being exported under the new global `Ember.Template`.

The `Ember.Template` is just a placeholder until the final global is chosen.

Serabe commented:
I redid the whole PR because this rebase was going to be harder than redoing everything.

Serabe commented:
Green again.

Serabe commented:
Closing this. Sorry, I cannot keep rebasing this.

locks commented:
Sorry 😢I'll wait for https://github.com/emberjs/ember.js/pull/16530.
